File: iosApp/iosApp/Navigator.swift
//
//  Navigator.swift
//  iosApp
//
//  Created by Vishnu Rajeevan on 12/13/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import shared

func navigator(component: AppComponent) -> some View{
  typealias ObservableNavigator = ObservableStore<NavigationStore, NavigationModel, NavigationEvents, NavigationEffects>
  let obsStore: ObservableNavigator
  
  let store = component.navigationStore
  let effectHandler = component.navigationEffectHandler
  obsStore = ObservableNavigator(
    store: store,
    state: NavigationModel.Companion.shared.defaultModel(),
    stateWatcher: store.watchState(),
    sideEffectWatcher: store.watchSideEffect()
  )
  
  _ = Loop<NavigationModel, NavigationEvents, NavigationEffects, NavigationEffectHandler>(
    store: store,
    effectHandler: effectHandler,
    startEffects: { Set() },
    eventSources: [Flow](),
    logTag: nil
  )
  
  return Navigator(component: component).environmentObject(obsStore)
}

struct Navigator: ConnectedView {
  typealias S = NavigationStore
  typealias M = NavigationModel
  typealias E = NavigationEvents
  typealias F = NavigationEffects
  
  let component: AppComponent
  
  struct Props {
    let navStack: [ScreenTarget]
    let navigationDispatch: (ScreenTarget) -> Void
  }
  
  
  func map(state: NavigationModel, dispatch: @escaping DispatchFunction<NavigationEvents>) -> Props {
    return Props(
      navStack: state.navStack,
      navigationDispatch: { target in dispatch(NavigationEvents.NavigateTo(target: target)) }
    )
  }
  
  
  func body(props: Props) -> some View {
    NavigationView{
        switch props.navStack.last {
        case ScreenTarget.root:
          LazyView { rootScreen(navigationDispatch: props.navigationDispatch) }
        case ScreenTarget.cars:
          LazyView { carScreen(component: component) }
        case ScreenTarget.tracks:
          LazyView { rootScreen(navigationDispatch: props.navigationDispatch) }
        case .none, .some(_):
          Text("Error")
        }
    }
  }
  
}

File: androidApp/src/main/java/me/vishnu/tracktracker/android/navigator/navigator.kt
package me.vishnu.tracktracker.android.navigator

import NavigationEvents
import NavigationModel
import ScreenTarget
import android.util.Log
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisallowComposableCalls
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.remember
import kotlinx.coroutines.flow.Flow
import me.vishnu.tracktracker.android.screens.CarScreen
import me.vishnu.tracktracker.android.screens.RootScreen
import me.vishnu.tracktracker.android.screens.TrackScreen
import me.vishnu.tracktracker.shared.graph.AppComponent
import me.vishnu.tracktracker.shared.graph.CarScreenComponent
import me.vishnu.tracktracker.shared.graph.TrackScreenComponent
import me.vishnu.tracktracker.shared.graph.create
import me.vishnu.tracktracker.shared.stores.Loop


@Composable
fun Navigator(component: AppComponent, backButton: Flow<NavigationEvents.GoBack>) {
  val store = component.navigationStore
  val effectHandler = component.navigationEffectHandler

  val (loopState, dispatch) = remember {
    Loop(
      store = store,
      effectHandler = effectHandler,
      eventSources = listOf(backButton),
      logTag = "Navigator"
    )
  }

  Navigator(
    component,
    loopState.collectAsState().value,
    dispatch
  )
}

@Composable
fun Navigator(
  appComponent: AppComponent,
  model: NavigationModel,
  dispatch: (NavigationEvents) -> Unit
) {
  val component = remember { appComponent }
  when (model.navStack.last()) {
    ScreenTarget.ROOT -> RootScreen(dispatch = dispatch)
    ScreenTarget.TRACKS -> {
      val trackScreenComponent = remember { TrackScreenComponent::class.create(component) }
      TrackScreen(
        store = trackScreenComponent.trackScreenStore,
        effectHandler = trackScreenComponent.trackScreenEffectHandler
      )
    }
    ScreenTarget.CARS -> {
      val carScreenComponent = remember { CarScreenComponent::class.create(component) }
      CarScreen(
        carScreenStore = carScreenComponent.carScreenStore,
        carScreenEffectHandler = carScreenComponent.carScreenEffectHandler
      )
    }
  }
}